thomasgauvin: add docs for hyperdrive to connect to private databases (#16766)
cloudflare/cloudflare-docs: src/content/docs/hyperdrive/configuration/connect-to-private-database.mdx

---
pcx_content_type: concept
title: Connect to a private database using Tunnel
sidebar:
  order: 4
---

import { TabItem, Tabs, Render } from "~/components";

Hyperdrive can securely connect to your private databases using [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) and [Cloudflare Access](/cloudflare-one/policies/access/).

## How it works

When your database is isolated within a private network (such as a [virtual private cloud](https://www.cloudflare.com/learning/cloud/what-is-a-virtual-private-cloud) or an on-premise network), you must enable a secure connection from your network to Cloudflare.

- [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) is used to establish the secure tunnel connection.
- [Cloudflare Access](/cloudflare-one/policies/access/) is used to restrict access to your tunnel such that only specific Hyperdrive configurations can access it.

<Render file="tutorials-before-you-start" product="workers" />

> ## Prerequisites:

Reviewer: ```suggestion
## Prerequisites
```

- A database in your private network, [configured to use TLS/SSL](/hyperdrive/configuration/connect-to-postgres/#supported-tls-ssl-modes).
- A hostname on your Cloudflare account, which will be used to route requests to your database.

## 1. Create a tunnel in your private network

### 1.1. Create a tunnel

First, create a [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) in your private network to establish a secure connection between your network and Cloudflare. Your network must be configured such that the tunnel has permissions to egress to the Cloudflare network and access the database within your network.

<Render file="tunnel/create-tunnel" product="cloudflare-one" />

### 1.2. Connect your database using a public hostname

Your tunnel must be configured to use a public hostname so that Hyperdrive can route requests to it.

1. In the **Public Hostnames** tab, choose a **Domain** and specify any subdomain or path information. This will be used in your Hyperdrive configuration to route to this tunnel.

> 2. In the **Service** section, specify **Type** `TCP` and the URL of your database, such as `127.0.0.1:5432`.

Author: ```suggestion
2. In the **Service** section, specify **Type** `TCP` and the URL of your database, such as `localhost:5432`. This must be the address and port of the database accessible from the machine on which the Tunnel is running.
```

> 2. In the **Service** section, specify **Type** `TCP` and the URL of your database, such as `127.0.0.1:5432`.

Reviewer 2: Should emphasize here that this is the egress end of the tunnel, and should point to whatever port the DB is configured to listen on.

Author: done!

> 2. In the **Service** section, specify **Type** `TCP` and the URL and configured port of your database, such as `localhost:5432`. 

Author: ```suggestion
2. In the **Service** section, specify **Type** `TCP` and the URL and configured port of your database, such as `localhost:5432`. This address will be used by the tunnel to route requests to your database.
```

3. Select **Save tunnel**.

## 2. Create a service token

The service token will be used to restrict requests to the tunnel, and is needed for the next step.

1. In [Zero Trust](https://one.dash.cloudflare.com), go to **Access** > **Service Auth** > **Service Tokens**.

2. Select **Create Service Token**.

3. Name the service token. The name allows you to easily identify events related to the token in the logs and to revoke the token individually.

> 4. Set a **Service Token Duration** of `Non-expiring`. This prevents the service token from expiring to be used throughout the life of the Hyperdrive configuration.

Reviewer: ```suggestion
4. Set a **Service Token Duration** of `Non-expiring`. This prevents the service token from expiring, ensuring it can be used throughout the life of the Hyperdrive configuration.
```

5. Select **Generate token**. You will see the generated Client ID and Client Secret for the service token, as well as their respective request headers.

6. Copy the Access Client ID and Access Client Secret. These will be used when creating the Hyperdrive configuration.

   :::caution

>    This is the only time Cloudflare Access will display the Client Secret. If you lose the Client Secret, you must generate a new service token.

Reviewer 2: I'd say instead "rotate" or "regenerate" the token, as it's still the same CF object.  The distinction matters for e.g. terraform.

Author: done!

>    This is the only time Cloudflare Access will display the Client Secret. If you lose the Client Secret, you must generate a new service token.

Author: ```suggestion
   This is the only time Cloudflare Access will display the Client Secret. If you lose the Client Secret, you must regenerate the service token.
```

   :::

## 3. Create an Access application to secure the tunnel

[Cloudflare Access](/cloudflare-one/policies/access/) will be used to verify that requests to the tunnel originate from Hyperdrive using the service token created above.

1. In [Zero Trust](https://one.dash.cloudflare.com), go to **Access** > **Applications**.

2. Select **Add an application**.

3. Select **Self-hosted**.

4. In **Application Configuration** > **Application name**, enter any name for the application.

5. In **Application Configuration** > **Session Duration**, select `No duration, expires immediately`.

6. In **Application Configuration** > **Application domain**, enter the subdomain and domain that was previously set for the tunnel application.

7. In **Application Appearance**, disable the `Enable App in App Launcher` setting.

> 8. In **Identify providers**, disable the `Accept all available identify providers` setting and select `Deselect all` identify providers.

Reviewer 3: ```suggestion
8. In **Identity providers**, disable the `Accept all available identity providers` setting and select `Deselect all` identity providers.
```

9. Select **Next**.

10. Enter a name in the **Policy name** and set the **Action** to `Service Auth`.

> 11. In **Configure rules**, create an **Include** rule. Specify a **Selector** of `Service Token` and the **Value** of the service token you created in step [2. Create a service token](#2-create-a-service-token).

Reviewer 2: Selector is not necessary here, just need the Include rule.

Author: I removed the Require rule, only keeping Include rule, configured with Selector > Service Token and Value > service token created above

12. Select **Next**.

13. Select **Add application** to create the Access application.

## 4. Use the Wrangler CLI to create a Hyperdrive configuration

> The Wrangler CLI will be used to create a Hyperdrive configuration using the Access application and Tunnel created above.

Author: Is it the same case for Access? in the same line

> The Wrangler CLI will be used to create a Hyperdrive configuration using the Access application and Tunnel created above.

Reviewer: I checked, but Cloudflare Access is sometimes simply referred to as "Access" (without "Cloudflare" in front).

```sh
# wrangler v3.65 and above required
npx wrangler hyperdrive create <NAME-OF-HYPERDRIVE-CONFIGURATION-FOR-DB-VIA-TUNNEL> --host=<HOSTNAME-FOR-THE-TUNNEL> --user=<USERNAME-FOR-YOUR-DATABASE> --password=<PASSWORD-FOR-YOUR-DATABASE> --database=<DATABASE-TO-CONNECT-TO> --access-client-id=<YOUR-ACCESS-CLIENT-ID> --access-client-secret=<YOUR-SERVICE-TOKEN-CLIENT-SECRET>
```

This command will create a Hyperdrive configuration using the usual database information (database name, database host, database user, and database password).

In addition, it will also set the Access Client ID and the Access Client Secret of the Service Token. When Hyperdrive makes requests to the tunnel, requests will be intercepted by Access and validated using the credentials of the Service Token.

## 5. Query your Hyperdrive configuration from a Worker (optional)

> To test your Hyperdrive configuration to the database using Tunnel and Access, use the Hyperdrive configuration ID in your Worker and deploy it.

Reviewer: ```suggestion
To test your Hyperdrive configuration to the database using Cloudflare Tunnel and Access, use the Hyperdrive configuration ID in your Worker and deploy it.
```

---
pcx_content_type: concept
title: Connect to a private database using Tunnel
sidebar:
  order: 4
---

import { TabItem, Tabs, Render } from "~/components";

Hyperdrive can securely connect to your private databases using [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) and [Cloudflare Access](/cloudflare-one/policies/access/).

## How it works

When your database is isolated within a private network (such as a [virtual private cloud](https://www.cloudflare.com/learning/cloud/what-is-a-virtual-private-cloud) or an on-premise network), you must enable a secure connection from your network to Cloudflare.

- [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) is used to establish the secure tunnel connection.
- [Cloudflare Access](/cloudflare-one/policies/access/) is used to restrict access to your tunnel such that only specific Hyperdrive configurations can access it.

<Render file="tutorials-before-you-start" product="workers" />

## Prerequisites:

- A database in your private network.

> - A default hostname on your Cloudflare account, which will be used to route requests to your database.

Reviewer 2: Any hostname would work, doesn't need to be a default?  Maybe "At least one domain on your Cloudflare account"?

> ## 1. Create a tunnel in your private network

Reviewer 4: nit: I prefer not using numbers here and just relying on the Markdown heading levels.

Reviewer: @OilyLime Thanks for your feedback! :D

I think this is alright to keep as-is for now, as our other tutorials often use numbered headings - it just makes it a bit clearer when users step through the tasks.

### 1.1. Create a tunnel

First, create a [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) in your private network to establish a secure connection between your network and Cloudflare. Your network must be configured such that the tunnel has permissions to egress to the Cloudflare network and access the database within your network.

<Render file="tunnel/create-tunnel" product="cloudflare-one" />

### 1.2. Connect your database using a public hostname

Your tunnel must be configured to use a public hostname so that Hyperdrive can route requests to it.

1. In the **Public Hostnames** tab, choose a **Domain** and specify any subdomain or path information. This will be used in your Hyperdrive configuration to route to this tunnel.

> 2. In the **Service** section, specify **Type** `TCP` and the URL of your database, such as `127.0.0.1:5432`.

Reviewer 4: Let's make sure we're consistent about `127.0.0.1` vs `localhost`

Reviewer 2: Would be good to double check whether 127.0.0.1 works, or if "localhost" is required.  Given it turned out to be the SSL issue, might be fine either way?

Author: Both work

1. In [Zero Trust](https://one.dash.cloudflare.com), go to **Access** > **Service Auth** > **Service Tokens**.

2. Select **Create Service Token**.

3. Name the service token. The name allows you to easily identify events related to the token in the logs and to revoke the token individually.

4. Set a **Service Token Duration** of `Non-expiring`. This prevents the service token from expiring to be used throughout the life of the Hyperdrive configuration.

5. Select **Generate token**. You will see the generated Client ID and Client Secret for the service token, as well as their respective request headers.

6. Copy the Access Client ID and Access Client Secret. These will be used when creating the Hyperdrive configuration.

   :::caution
   This is the only time Cloudflare Access will display the Client Secret. If you lose the Client Secret, you must generate a new service token.
   :::

## 3. Create an Access application to secure the tunnel

[Cloudflare Access](/cloudflare-one/policies/access/) will be used to verify that requests to the tunnel originate from Hyperdrive using the service token created above.

> 1. In [Zero Trust](https://one.dash.cloudflare.com), go to **Access** > **Applications**.

Reviewer 4: My only concern is that these instructions become obsolete if ZT changes their dashboard experience...

Author: I tried to re-use as much as possible. The issue lies in the fact that we need to provide opinionated guidance for some of these. For instance, we need to indicate that they need to specify Type TCP and something such as localhost:5432 when creating the tunnel.

Here, we need to specify the service auth token that needs to be configured.

13. Select **Add application** to create the Access application.

## 4. Create a Hyperdrive configuration

To create a Hyperdrive configuration for your private database, you'll need to specify the Access application and Tunnel information upon creation.

<Tabs> <TabItem label="Wrangler">

```sh
# wrangler v3.65 and above required
npx wrangler hyperdrive create <NAME-OF-HYPERDRIVE-CONFIGURATION-FOR-DB-VIA-TUNNEL> --host=<HOSTNAME-FOR-THE-TUNNEL> --user=<USERNAME-FOR-YOUR-DATABASE> --password=<PASSWORD-FOR-YOUR-DATABASE> --database=<DATABASE-TO-CONNECT-TO> --access-client-id=<YOUR-ACCESS-CLIENT-ID> --access-client-secret=<YOUR-SERVICE-TOKEN-CLIENT-SECRET>
```

</TabItem> <TabItem label="Terraform">

```terraform
resource "cloudflare_hyperdrive_config"  "<TERRAFORM_VARIABLE_NAME_FOR_CONFIGURATION>" {
  account_id = "<YOUR_ACCOUNT_ID>"
  name       = "<NAME_OF_HYPERDRIVE_CONFIGURATION>"
  origin     = {
    host     = "<HOSTNAME_OF_TUNNEL>"
    database = "<NAME_OF_DATABASE>"
    user     = "<NAME_OF_DATABASE_USER>"
    password = "<DATABASE_PASSWORD>"
    scheme   = "postgres"
    access_client_id     = "<ACCESS_CLIENT_ID>"
    access_client_secret = "<ACCESS_CLIENT_SECRET>"

>   }

Author: ```suggestion
  }
  caching = {
    disabled = false
  }
```

- [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) is used to establish the secure tunnel connection.
- [Cloudflare Access](/cloudflare-one/policies/access/) is used to restrict access to your tunnel such that only specific Hyperdrive configurations can access it.

<Render file="tutorials-before-you-start" product="workers" />

## Prerequisites

- A database in your private network, [configured to use TLS/SSL](/hyperdrive/configuration/connect-to-postgres/#supported-tls-ssl-modes).
- A hostname on your Cloudflare account, which will be used to route requests to your database.

## 1. Create a tunnel in your private network

### 1.1. Create a tunnel

First, create a [Cloudflare Tunnel](/cloudflare-one/connections/connect-networks/) in your private network to establish a secure connection between your network and Cloudflare. Your network must be configured such that the tunnel has permissions to egress to the Cloudflare network and access the database within your network.

<Render file="tunnel/create-tunnel" product="cloudflare-one" />

### 1.2. Connect your database using a public hostname

> Your tunnel must be configured to use a public hostname so that Hyperdrive can route requests to it.

Author: ```suggestion
Your tunnel must be configured to use a public hostname so that Hyperdrive can route requests to it. If you don't have a hostname on Cloudflare yet, you will need to [register a new hostname](/registrar/get-started/register-domain/) or [add a zone](https://developers.cloudflare.com/dns/zone-setups/) to proceed. 
```

> 1. In the **Public Hostnames** tab, choose a **Domain** and specify any subdomain or path information. This will be used in your Hyperdrive configuration to route to this tunnel.

Reviewer 5: I would add in here that if you don't have a domain on Cloudflare yet, you will need to register one or add a zone (bc i saw someone get stuck on that)

Author: thanks, adding it, it's in the prereqs that's might be getting glossed over

1. In [Zero Trust](https://one.dash.cloudflare.com), go to **Access** > **Applications**.

2. Select **Add an application**.

3. Select **Self-hosted**.

4. In **Application Configuration** > **Application name**, enter any name for the application.

5. In **Application Configuration** > **Session Duration**, select `No duration, expires immediately`.

6. In **Application Configuration** > **Application domain**, enter the subdomain and domain that was previously set for the tunnel application.

7. In **Application Appearance**, disable the `Enable App in App Launcher` setting.

8. In **Identity providers**, disable the `Accept all available identity providers` setting and select `Deselect all` identity providers.

9. Select **Next**.

10. Enter a name in the **Policy name** and set the **Action** to `Service Auth`.

11. In **Configure rules**, create an **Include** rule. Specify a **Selector** of `Service Token` and the **Value** of the service token you created in step [2. Create a service token](#2-create-a-service-token).

12. Select **Next**.

13. Select **Add application** to create the Access application.

## 4. Create a Hyperdrive configuration

> To create a Hyperdrive configuration for your private database, you'll need to specify the Access application and Tunnel information upon creation.

Author: ```suggestion
To create a Hyperdrive configuration for your private database, you'll need to specify the Access application and Cloudflare Tunnel information upon creation.
```

<Tabs> <TabItem label="Wrangler">

```sh
# wrangler v3.65 and above required
npx wrangler hyperdrive create <NAME-OF-HYPERDRIVE-CONFIGURATION-FOR-DB-VIA-TUNNEL> --host=<HOSTNAME-FOR-THE-TUNNEL> --user=<USERNAME-FOR-YOUR-DATABASE> --password=<PASSWORD-FOR-YOUR-DATABASE> --database=<DATABASE-TO-CONNECT-TO> --access-client-id=<YOUR-ACCESS-CLIENT-ID> --access-client-secret=<YOUR-SERVICE-TOKEN-CLIENT-SECRET>
```

</TabItem> <TabItem label="Terraform">

```terraform
resource "cloudflare_hyperdrive_config"  "<TERRAFORM_VARIABLE_NAME_FOR_CONFIGURATION>" {
  account_id = "<YOUR_ACCOUNT_ID>"
  name       = "<NAME_OF_HYPERDRIVE_CONFIGURATION>"
  origin     = {
    host     = "<HOSTNAME_OF_TUNNEL>"
    database = "<NAME_OF_DATABASE>"
    user     = "<NAME_OF_DATABASE_USER>"
    password = "<DATABASE_PASSWORD>"
    scheme   = "postgres"
    access_client_id     = "<ACCESS_CLIENT_ID>"
    access_client_secret = "<ACCESS_CLIENT_SECRET>"
  }
  caching = {
    disabled = false
  }
}
```

</TabItem> </Tabs>

> This command will create a Hyperdrive configuration using the usual database information (database name, database host, database user, and database password).

Author: ```suggestion
This will create a Hyperdrive configuration using the usual database information (database name, database host, database user, and database password).
```
